mesonbuild/utils/universal.py

```diff
@@ -1126,6 +1127,30 @@ def determine_worker_count(varnames: T.Optional[T.List[str]] = None) -> int:
             num_workers = 1
     return num_workers
 
+def is_parent_path(parent: str, trial: str) -> bool:
+    '''Checks if @trial is a file under the directory @parent. Both @trial and @parent should be
+       adequately normalized, though empty and '.' segments in @parent and @trial are accepted
+       and discarded, matching the behavior of os.path.commonpath.  Either both or none should
+       be absolute.'''
+    assert os.path.isabs(trial) == os.path.isabs(parent)
+    if is_windows():
+        parent = parent.replace('\\', '/')
+        trial = trial.replace('\\', '/')
+
+    split_parent = parent.split('/')
+    split_trial = trial.split('/')
+
+    split_parent = [c for c in split_parent if c and c != '.']
+    split_trial = [c for c in split_trial if c and c != '.']
+
+    if len(split_trial) < len(split_parent):
+        return False
+    for i, component in enumerate(split_parent):
+        if component != split_trial[i]:
+            return False
```

Comment:
I'm probably nitpicking, but my feeling would be to use zip instead:

```suggestion
    for parent_comp, trial_comp in zip(split_parent, split_trial):
        if parent_comp != trial_comp:
            return False
```

or even:

```
return all(pc == tc for pc, tc in zip(split_parent, split_trial)
```

Reply:
Cool, I would have sworn that zip returned an error if you iterate past the shortest argument; instead it stops.

Related to this I like in my version that it's more explicit how you only iterate on the parent's components; on the other hand `enumerate` is a bit of an antipattern so I can change it.

Reply:
Python 3.10 added a `strict` parameter, to error if lengths are different.

Reply:
A third option would be : `return tuple(split_parent) == tuple(split_trial[:len(split_parent)])`

Reply:
I agree with @bruchar1 latest suggestion, except that I don't think the cast to tuple is necessary. Not having the for-loop is also probably imperceptibly faster.

Reply:
You're right. I was wrongly thinking the equality operator was only working with tuples. Furthermore, it removes the need for the length comparison as the comparison would return False if the trial list is smaller.